File: app/src/main/java/com/eurobuddha/pandapools/ReAnnouncer.java
package com.eurobuddha.pandapools;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Layer 5 (network discoverability): keep a live announce beacon in the recent sentinel window for pools,
 * so STRANGERS' fresh nodes keep rediscovering them after the original beacon pruned.
 *
 * GOSSIP: a pool stays discoverable only while a live beacon sits in the recent window, and the beacon is
 * unspendable dust that prunes (~a day). If re-announcing were owner-only, a pool created on a now-offline
 * phone would go dark for everyone. So this re-announces EVERY funded pool the node knows (its own +
 * any it has discovered/tracked), turning discovery into a self-healing mesh: as long as one active node
 * knows a funded pool, its beacon stays alive for new users. Always-on nodes (desktop/MDS) are the anchors.
 * Re-announcing another's pool is safe — {@link PoolManager#reannounce} spends no covenant coin and needs
 * no owner signature; it just posts a dust beacon carrying the pool's public params (which the node holds).
 *
 * It re-announces ONLY a pool whose beacon has actually FADED from the recent window — so beacons never
 * pile up — and shuffles + caps each run so many nodes don't all re-beacon the same pools at once.
 *
 * HARD GUARD: the sentinel scan here NEVER uses {@code megammr:true} — that query (pulling the whole
 * accumulated beacon pile) was the 0.8.10 crash, not the re-announce itself. Recent unpruned window only.
 */
public class ReAnnouncer {

    public interface Listener { void onAnnounced(int posted); }

    /** Per-run cap on how many faded beacons ONE node re-posts. Gossip is collective — shuffle + cap so we
     *  don't all beacon the same pools at once; other nodes and later runs cover the rest. */
    private static final int MAX_PER_RUN = 8;

    // Keys re-announced during THIS process — so a config-change recreate (e.g. theme toggle) can't
    // double-post a beacon before the first one confirms into the sentinel window. RE-ARMED (key removed) in
    // announce() the moment that beacon shows up live in the recent window, so a later re-fade re-announces
    // again (an offline owner's pool survives multiple fade cycles). Cleared entirely on process death.
    private static final Set<String> ANNOUNCED = java.util.Collections.synchronizedSet(new HashSet<>());

    private final NodeApi node;
    private final PoolManager mgr;

    public ReAnnouncer(NodeApi node) {
        this.node = node;
        this.mgr = new PoolManager(node);
    }

    /**
     * For each currently-funded owned pool whose beacon is absent from the recent sentinel window, post one
     * fresh beacon. {@code fundedOwn} are pools with live reserves + full params (from discovery).
     */
    public void refreshFaded(final List<Pool> fundedOwn, final Listener cb) {
        if (fundedOwn == null || fundedOwn.isEmpty()) { cb.onAnnounced(0); return; }
        // PROACTIVE present-check: scan the TIGHTER REANNOUNCE_DEPTH window (not the wider discovery depth), so a
        // beacon counts as "faded" — and gets re-announced — once it is older than REANNOUNCE_DEPTH, i.e. BEFORE
        // it would leave the SENTINEL_SCAN_DEPTH discovery window. That's what keeps a live pool's beacon from ever
        // lapsing on a non-owner node (the cross-device flicker fix). NEVER megammr:true (see class note); still
        // depth-bounded so the reply stays IPC-safe.
        node.cmd("coins simplestate:true order:desc depth:" + PoolCovenant.REANNOUNCE_DEPTH
                + " address:" + PoolCovenant.SENTINEL, new NodeApi.Cb() {
            @Override public void onResult(JSONObject j) { announce(present(j), fundedOwn, cb); }
            @Override public void onError(String m) { cb.onAnnounced(0); }   // can't tell what's live → do nothing
        });
    }

    /**
     * Background entry (no UI state): build the funded owned-pool list straight from {@link OwnPoolStore}
     * — a reserve scan per recipe, no ownership/keys lookup needed (a stored recipe IS ours) — then
     * re-announce the faded ones. Skips recipes with no live reserves (a closed pool must not be announced).
     */
    public void refreshFadedFromStore(final Context ctx, final Listener cb) {
        final List<Pool> recipes = OwnPoolStore.all(ctx);
        if (recipes.isEmpty()) { cb.onAnnounced(0); return; }
        final List<Pool> funded = Collections.synchronizedList(new ArrayList<>());
        final AtomicInteger pending = new AtomicInteger(recipes.size());
        for (final Pool r : recipes) {
            node.cmd("coins address:" + r.address, new NodeApi.Cb() {
                @Override public void onResult(JSONObject j) { fillReserves(r, j); if (r.funded()) funded.add(r); tick(); }
                @Override public void onError(String m) { tick(); }
                private void tick() { if (pending.decrementAndGet() == 0) refreshFaded(new ArrayList<>(funded), cb); }
            });
        }
    }

    /**
     * GOSSIP entry (background + foreground): run a full registry scan (Source-1 OwnPoolStore ∪ Source-2 live
     * beacons), then re-announce EVERY funded pool whose beacon has faded — not just this node's own. This
     * is the self-healing mesh: any node that has discovered a pool helps keep it alive, so a pool created
     * on a now-offline phone stays findable while it holds reserves and at least one active node knows it.
     * {@link PoolBook#scan} returns the funded, deduped pool set; {@link #refreshFaded} filters to faded.
     */
    public void refreshFadedFromScan(final Listener cb) {
        new PoolBook(node).scan(new PoolBook.Listener() {
            @Override public void onPools(List<Pool> funded) { refreshFaded(funded, cb); }
            @Override public void onError(String msg) { cb.onAnnounced(0); }
        });
    }

    /** Largest coin per leg at the covenant address = the true reserve (dust can't masquerade). Mirrors
     *  PoolBook.fund so a store-built pool matches a discovery-built one. */
    private static void fillReserves(Pool p, JSONObject j) {
        Object resp = j.opt("response");
        JSONArray cs = resp instanceof JSONArray ? (JSONArray) resp : new JSONArray();
        for (int i = 0; i < cs.length(); i++) {
            JSONObject c = cs.optJSONObject(i);
            if (c == null || c.optBoolean("spent", false)) continue;
            String tid = c.optString("tokenid", "");
            if ("0x00".equals(tid)) {
                BigDecimal amt = new BigDecimal(c.optString("amount", "0"));
                if (p.reserveM == null || amt.compareTo(p.reserveM) > 0) { p.reserveM = amt; p.coinidM = c.optString("coinid", ""); }
            } else if (p.tok != null && p.tok.equalsIgnoreCase(tid)) {
                BigDecimal amt = new BigDecimal(c.optString("tokenamount", c.optString("amount", "0")));
                if (p.reserveT == null || amt.compareTo(p.reserveT) > 0) { p.reserveT = amt; p.coinidT = c.optString("coinid", ""); }
            }
        }
    }

    /** The set of pool keys (opk|tok|kmin) that currently have a live beacon in the recent window. */
    private static Set<String> present(JSONObject j) {
        Set<String> keys = new HashSet<>();
        Object resp = j.opt("response");
        JSONArray coins = resp instanceof JSONArray ? (JSONArray) resp : new JSONArray();
        for (int i = 0; i < coins.length(); i++) {
            JSONObject c = coins.optJSONObject(i);
            if (c == null) continue;
            // full beacon identity incl. oadr (port 3) — a forged beacon with your public opk/tok/kmin but a
            // bogus oadr must NOT be mistaken for your live beacon (it would suppress your re-announce and
            // point fresh nodes at a wrong-address covenant). Strict parity with PoolBook.gatherRegistry.
            String tok = PoolBook.state(c, 2), oadr = PoolBook.state(c, 3), opk = PoolBook.state(c, 4), kmin = PoolBook.state(c, 5);
            if (tok != null && oadr != null && opk != null && kmin != null) keys.add(key(opk, oadr, tok, kmin));
        }
        return keys;
    }

    private void announce(Set<String> present, List<Pool> fundedOwn, final Listener cb) {
        List<Pool> faded = new ArrayList<>();
        for (Pool p : fundedOwn) {
            if (p == null || p.opk == null || p.oadr == null || p.tok == null || p.kmin == null) continue;
            String k = key(p.opk, p.oadr, p.tok, p.kmin);
            if (present.contains(k)) {
                // Beacon is live in the recent window again → RE-ARM: drop the process-dedup guard so that when this
                // beacon fades AGAIN (~20h later) we re-announce it, keeping an offline owner's pool alive across
                // multiple fade cycles for the whole life of this (long-lived) foreground process. Without this the
                // static set is sticky and a discovered pool is re-announced at most ONCE per process, letting its
                // beacon go permanently dark on a later re-fade. Mirrors MDS service.js `delete ANN_SVC[abk]`.
                ANNOUNCED.remove(k);
                continue;
            }
            if (ANNOUNCED.contains(k)) continue;   // already re-posted this process, beacon not yet confirmed into window
            faded.add(p);
        }
        if (faded.isEmpty()) { cb.onAnnounced(0); return; }

        // Gossip is collective: shuffle + cap per run so many nodes don't all re-beacon the same faded pools
        // at once. Whatever this run skips, another node or a later run covers.
        if (faded.size() > MAX_PER_RUN) {
            Collections.shuffle(faded);
            faded = new ArrayList<>(faded.subList(0, MAX_PER_RUN));
        }

        final AtomicInteger pending = new AtomicInteger(faded.size());
        final AtomicInteger posted = new AtomicInteger(0);
        for (Pool p : faded) {
            final String k = key(p.opk, p.oadr, p.tok, p.kmin);
            mgr.reannounce(p, new PoolManager.Result() {
                @Override public void onPosted(String txpowid) { ANNOUNCED.add(k); posted.incrementAndGet(); tick(); }
                @Override public void onFailed(String message) { tick(); }
                private void tick() { if (pending.decrementAndGet() == 0) cb.onAnnounced(posted.get()); }
            });
        }
    }

    private static String key(String opk, String oadr, String tok, String kmin) {
        return (opk + "|" + oadr + "|" + tok + "|" + kmin).toLowerCase();
    }
}

File: anchor/reannouncer.py
#!/usr/bin/env python3
"""
PandaPools cloud ANCHOR re-announcer.

Keeps every funded pool's registry beacon fresh so cross-node discovery never lapses. Runs on an always-on node
(the eurobuddha megammr node @ 127.0.0.1:9005) on a timer. The phones' discovery only finds OTHER creators' pools
via live beacons in the recent sentinel window, and a beacon is unspendable dust that MMR-prunes after ~a day —
so if nobody re-announces, the mesh collapses (each phone falls back to only its own tracked pools). This process
is the anchor the app's ReAnnouncer comment envisions: as long as it runs, every funded pool stays discoverable.

PROACTIVE (not reactive): re-announces a pool whose newest beacon is older than STALE_BLOCKS (~12h), i.e. BEFORE
the ~24h prune — no dark window.

FUND-SAFETY: read-only except a dust `send` of 0.000000001 MINIMA to the UNSPENDABLE sentinel, funded from the
node wallet, change back. It NEVER spends a covenant/pool coin and NEVER signs an owner key (the beacon just
carries the pool's public params, exactly like PoolManager.reannounce / the MDS poolmgr.reannounce). It re-posts
ONLY pools that are (a) on-chain funded and (b) whose 0.5% covenant `parseok`s — a dead/closed/unparseable pool is
never re-announced.

Config via env: PP_RPC (default http://127.0.0.1:9005/), PP_FAUCET, PP_STALE_BLOCKS, PP_MIN_SENDABLE,
PP_MAX_PER_RUN, PP_DRY=1. CLI: --dry-run (log what it WOULD post, no send).
"""
import json, urllib.request, urllib.parse, sys, os, time
from decimal import Decimal

RPC          = os.environ.get("PP_RPC", "http://127.0.0.1:9005/")
FAUCET       = os.environ.get("PP_FAUCET", "https://eurobuddha.com/faucet/api/request")
SENTINEL     = "0x50414E4441504F4F4C53"                 # "PANDAPOOLS"
MXUSDT       = "0x7D39745FBD29049BE29850B55A18BF550E4D442F930F86266E34193D89042A90"
DUST         = "0.000000001"
STALE_BLOCKS = int(os.environ.get("PP_STALE_BLOCKS", "900"))       # ~12h @ ~50s/block — refresh before the ~24h prune
MIN_SENDABLE = Decimal(os.environ.get("PP_MIN_SENDABLE", "0.001")) # below this, top up from the faucet
MAX_PER_RUN  = int(os.environ.get("PP_MAX_PER_RUN", "40"))
DRY          = ("--dry-run" in sys.argv) or (os.environ.get("PP_DRY") == "1")

# The LIVE 0.5% covenant — byte-identical to PoolCovenant.TEMPLATE (native) and covenant.js (MDS). NOT the 0.3%
# harness template: address = hash(script), so a 0.3% reconstruction would derive an empty address and every real
# pool would look "unfunded". The KMIN literal is taken verbatim from the beacon (already canonical).
TEMPLATE = (
 "IF SIGNEDBY($OPK) THEN "
 "IF VERIFYOUT(@INPUT $OADR @AMOUNT @TOKENID FALSE) THEN RETURN TRUE ENDIF "
 "RETURN GETOUTADDR(@INPUT) EQ @ADDRESS AND GETOUTTOK(@INPUT) EQ @TOKENID AND GETOUTAMT(@INPUT) GTE @AMOUNT "
 "ENDIF "
 "IF @TOKENID EQ 0x00 THEN "
 "ASSERT @INPUT % 2 EQ 0 LET s=@INPUT+1 "
 "ASSERT GETINADDR(s) EQ @ADDRESS AND GETINTOK(s) EQ $TOK "
 "ASSERT GETOUTADDR(s) EQ @ADDRESS AND GETOUTTOK(s) EQ $TOK "
 "LET x=@AMOUNT LET y=GETINAMT(s) LET nx=GETOUTAMT(@INPUT) LET ny=GETOUTAMT(s) "
 "ASSERT VERIFYOUT(@INPUT @ADDRESS nx 0x00 FALSE) "
 "ELSE "
 "ASSERT @TOKENID EQ $TOK AND @INPUT % 2 EQ 1 LET s=@INPUT-1 "
 "ASSERT GETINADDR(s) EQ @ADDRESS AND GETINTOK(s) EQ 0x00 "
 "ASSERT GETOUTADDR(s) EQ @ADDRESS AND GETOUTTOK(s) EQ 0x00 "
 "LET y=@AMOUNT LET x=GETINAMT(s) LET ny=GETOUTAMT(@INPUT) LET nx=GETOUTAMT(s) "
 "ASSERT VERIFYOUT(@INPUT @ADDRESS ny $TOK FALSE) "
 "ENDIF "
 "LET dx=nx-x LET dy=ny-y LET fx=MAX(dx 0)*5/1000 LET fy=MAX(dy 0)*5/1000 "
 "RETURN (nx-fx)*(ny-fy) GTE MAX(x*y $KMIN)")


def rpc(cmd, timeout=180):
    with urllib.request.urlopen(RPC + urllib.parse.quote(cmd, safe=""), timeout=timeout) as r:
        return json.load(r)


def resp_list(r):
    x = r.get("response")
    return x if isinstance(x, list) else []


def block():
    return int(rpc("block")["response"]["block"])


def smap(coin):
    """simplestate map {port: value} for a coin (already simplestate:true)."""
    s = coin.get("state")
    return s if isinstance(s, dict) else {}


def pool_script(opk, oadr, tok, kmin):
    return TEMPLATE.replace("$OPK", opk).replace("$OADR", oadr).replace("$TOK", tok).replace("$KMIN", kmin)


def derive(opk, oadr, tok, kmin):
    """runscript the reconstructed 0.5% covenant → (parseok, address). json.dumps quotes the script and does NOT
       escape '/', so '*5/1000' stays a real slash (the '\\/' bug that strands funds is impossible here)."""
    try:
        r = rpc("runscript script:" + json.dumps(pool_script(opk, oadr, tok, kmin)))
        resp = r.get("response") or {}
        pj = resp.get("parseok")
        parseok = (pj is True) or (str(pj).lower() == "true")
        addr = ((resp.get("script") or {}).get("address")) if parseok else None
        return parseok, addr
    except Exception as e:
        return False, None


def reserves(address, tok):
    """Largest coin per leg = the true reserve (dust can't masquerade). Returns (reserveM, reserveT).
       megammr:true so an anchor that doesn't TRACK the pool covenant still sees its reserves (bounded: ~2 coins
       at this one address, not the sentinel pile — safe on a server)."""
    rM = rT = Decimal(0)
    for c in resp_list(rpc("coins address:" + address + " megammr:true")):
        if c.get("spent"):
            continue
        tid = c.get("tokenid", "")
        if tid == "0x00":
            a = Decimal(str(c.get("amount", "0")))
            if a > rM: rM = a
        elif tid.lower() == tok.lower():
            a = Decimal(str(c.get("tokenamount", c.get("amount", "0"))))
            if a > rT: rT = a
    return rM, rT


def sendable_minima():
    try:
        b = rpc("balance tokenid:0x00")["response"]
        row = b[0] if isinstance(b, list) and b else {}
        return Decimal(str(row.get("sendable", "0")))
    except Exception:
        return Decimal(0)


def faucet_topup():
    """Best-effort: request a drip from the co-located faucet so the anchor self-funds. Never raises."""
    try:
        addr = rpc("getaddress")["response"].get("address", "")
        if not addr:
            return
        url = FAUCET + "?address=" + urllib.parse.quote(addr, safe="")
        with urllib.request.urlopen(url, timeout=30) as r:
            body = r.read(2048).decode("utf-8", "replace")
        print("  [fund] faucet requested for", addr[:14], "->", body[:120])
    except Exception as e:
        print("  [fund] faucet top-up failed:", e)


def reannounce(state_map):
    """Relay a fresh beacon: dust send to the sentinel carrying the SAME state map (verbatim, no reconstruction)."""
    st = json.dumps({k: state_map[k] for k in sorted(state_map, key=lambda x: int(x))}, separators=(",", ":"))
    cmd = "send amount:%s address:%s tokenid:0x00 state:%s" % (DUST, SENTINEL, st)
    r = rpc(cmd)
    if not (r.get("status") is True or r.get("pending") is True):
        raise RuntimeError(str(r.get("error") or r.get("response"))[:160])
    return (r.get("response") or {}).get("txpowid", "")


def main():
    tip = block()
    beacons = resp_list(rpc("coins simplestate:true address:%s megammr:true" % SENTINEL))
    print("run @ tip %d | beacons on chain: %d | dry_run=%s | stale>%d blk (~%.0fh)" %
          (tip, len(beacons), DRY, STALE_BLOCKS, STALE_BLOCKS * 50 / 3600))

    # group by pool key opk|tok|kmin, keep the newest beacon's state + created block
    pools = {}
    for c in beacons:
        s = smap(c)
        tok, oadr, opk, kmin = s.get("2"), s.get("3"), s.get("4"), s.get("5")
        if not (tok and oadr and opk and kmin):
            continue
        created = int(c.get("created", "0") or 0)
        key = (opk + "|" + tok + "|" + kmin).lower()
        cur = pools.get(key)
        if cur is None or created > cur["created"]:
            pools[key] = {"opk": opk, "oadr": oadr, "tok": tok, "kmin": kmin, "created": created, "state": s}

    scanned = len(pools)
    funded = faded = posted = 0
    to_post = []
    for p in pools.values():
        age = tip - p["created"]
        parseok, addr = derive(p["opk"], p["oadr"], p["tok"], p["kmin"])
        if not parseok or not addr:
            continue
        rM, rT = reserves(addr, p["tok"])
        is_funded = rM > 0 and rT > 0
        if is_funded:
            funded += 1
        stale = age > STALE_BLOCKS
        tag = "FUNDED" if is_funded else "empty "
        print("  %s age %5d blk (~%4.1fh) %s M=%s T=%s  %s" %
              (tag, age, age * 50 / 3600, addr[:14] + "…", rM, rT,
               ("-> RE-ANNOUNCE" if (is_funded and stale) else ("fresh" if is_funded else "skip"))))
        if is_funded and stale:
            faded += 1
            to_post.append(p)

    to_post = to_post[:MAX_PER_RUN]
    if to_post and not DRY:
        if sendable_minima() < MIN_SENDABLE:
            faucet_topup()
            time.sleep(2)
        for i, p in enumerate(to_post):
            if i:
                time.sleep(3)   # small gap so consecutive sends pick DISTINCT coins (avoid same-UTXO race)
            try:
                tx = reannounce(p["state"])
                posted += 1
                print("  [post] re-announced %s… tx %s" % (p["state"].get("4", "?")[:14], tx[:16]))
            except Exception as e:
                print("  [post] FAILED %s… : %s" % (p["state"].get("4", "?")[:14], e))
                if "funds" in str(e).lower() or "coin" in str(e).lower():
                    faucet_topup(); time.sleep(2)

    print("SUMMARY scanned=%d funded=%d faded=%d posted=%d%s" %
          (scanned, funded, faded, posted, " (DRY-RUN, nothing sent)" if DRY else ""))


if __name__ == "__main__":
    main()
